### ISSUE TITLE
Garbage text shown in recording information

### ISSUE THREAD

**Issue by Tyri0n** [betonme/e2openplugin-EnhancedMovieCenter#289]
Just started using EMC and half of my recordings' info is displaying garbage. The text is displayed correctly in the default video playlist, and elsewhere before recording.

Looks like the issue is here:

20220113 23:25 [META] Detected short_event encoding-type: windows-1253 (0.330670671343)
20220113 23:25 [META] Exception in readEitFile: 'charmap' codec can't decode byte 0x9f in position 15: character maps to <undefined>
Traceback (most recent call last):
  File "/usr/lib/enigma2/python/Plugins/Extensions/EnhancedMovieCenter/EitSupport.py", line 416, in __readEitFile
  File "/usr/lib/python2.7/encodings/cp1253.py", line 15, in decode
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9f in position 15: character maps to <undefined>

**Comment by wedebe** [comment]
I think this is caused by EPGCache which provides descriptions for recordings.

But I also think that data from sources outside EMC should be sanitised by EMC upon ingestion.

**Comment by pjsharp** [comment]
@Tyri0n Could you please attach the eit file here, if you still have this problem?

**Comment by wedebe** [comment]
Sample attached:
(see also https://github.com/openatv/enigma2/issues/1954)

[Archive.zip](https://github.com/betonme/e2openplugin-EnhancedMovieCenter/files/10254511/Archive.zip)

![1_0_1_207A_808_2_11A0000_0_0_0_20221218194603](https://user-images.githubusercontent.com/9741693/208316292-179ad2d6-7df3-4eb4-90e4-9df70b9e79f8.jpg)

**Comment by Tyri0n** [comment]
I went back to the standard movie player soon after. The reply by @wedebe is what I was seeing, if I remember correctly.

**Comment by pjsharp** [comment]
A few days ago I opened a pull request that should solve these problems in EMC. It's still waiting to be merged.

![encoding2](https://user-images.githubusercontent.com/70219004/208318710-5f5fbad0-bc7d-46d4-a215-9cbf2fee9c7e.PNG)

**Comment by wedebe** [comment]
@Tyri0n, I've manually uploaded the updated files in your PR to my receiver - looks good :) Thanks a mill!

Does this mean that EPGCache wasn't actually at fault or what exactly was going on?

**Comment by Tyri0n** [comment]
> @Tyri0n, I've manually uploaded the updated files in your PR to my receiver - looks good :) Thanks a mill!
> 
> Does this mean that EPGCache wasn't actually at fault or what exactly was going on?
> 
> ![4097_0_2_0_0_0_0_0_0_0_20221219135719](https://user-images.githubusercontent.com/9741693/208442813-00d751c3-8a93-4bfd-83bc-6c3d09e76a20.jpg)

guessing this was meant for @pjsharp :)

**Comment by wedebe** [comment]
Sorry Tyri0n, yes it was 🙈

**Comment by pjsharp** [comment]
The eit-Parser of EMC did not support some encodings. I've added a few encoding and fixed some parsing errors I noticed.

#298

**Comment by dazulrich** [comment]
@wedebe I'm pretty sure that oatv epgcache (or whatever code) is at fault. I have same garbage (also recording 28.2e). I can make 2 short recordings from same program and 1 has garbage chars in eit file and the other is normal... It's great that pjsharp has added the decoding I'm emc, but it will make it less likely that oatv will try fixing the underlying problem..

**Comment by Tyri0n** [comment]
@dazulrich wouldn't I see garbage in the standard oatv movie player, in that case?

**Comment by pjsharp** [comment]
@dazulrich Do you still have these 2 recordings from the same program? Or would it be possible to reproduce it again? I would like to inspect the eit files.

**Comment by wedebe** [comment]
> @dazulrich wouldn't I see garbage in the standard oatv movie player, in that case?

Not if the standard oatv movie player has a workaround in it.

**Comment by wedebe** [comment]
> It's great that pjsharp has added the decoding I'm emc, but it will make it less likely that oatv will try fixing the underlying problem..

I raised this exact same concern with a workaround that got added to a different plugin.

Although, self-conflictingly... I also believe that any data consumer should never blindly trust an external data provider.

But, then againnnnnn... each individual workaround is another element added to a house of cards lined up for crapping itself if/when the original cause of such a problem is fixed at source 🙈

**Comment by wedebe** [comment]
> @dazulrich Do you still have these 2 recordings from the same program? Or would it be possible to reproduce it again? I would like to inspect the eit files.

I'd observed this too, but could never figure out a pattern. I'm sure I've got plenty of samples, just not available to me this week.

**Comment by pjsharp** [comment]
@wedebe What ever problems still exist in the background, adding support for more encodings to the plugin is not a workaround.  The information in an EIT file can be encoded with different encodings. Even within a file, individual event parts can be encoded differently (eg. name/title in utf-8, short description iso-8859-1 etc.) This is defined in the ETSI EN 300 468.

The EIT parser by emc plugin did not fully evaluate the information about the encoding used. Therefore, the correct encoding could not always be recognized although it could have been. If the encoding was not recognized, the content was subjected to an automatic encoding recognition. However, this also could not always recognize the encoding correctly. Then it happened that some characters were not displayed correctly.

In your case it is a special encoding - freesat huffman. This can also be seen in the EIT file (0x1F 0x02). What you call "garbage characters" is just huffman compressed text. And to bring it to the normal form I added the decoder to the plugin, because python can't easily decode it like utf-8 or other iso encodings.

I just extended the parser so that it can correctly recognize several encodings. No workaround!

**Comment by dazulrich** [comment]
Hi,
did some recordings.. 
the first one contains garbage.
the 2nd one is fine.

Between the recordings I opened 2nd infobar to display description.
have a look at the meta files. the description is slightly different. the 1st one has extra [S] at the end.
I copied the epg.dat and saved it again when the recording was working. 
in the original epg , i can only find the huffman encoded text, in the second one I can find both the encoded and the clear text description. 
[blue bloods.zip](https://github.com/betonme/e2openplugin-EnhancedMovieCenter/files/10291428/blue.bloods.zip)

**Comment by dazulrich** [comment]
> @dazulrich wouldn't I see garbage in the standard oatv movie player, in that case?

maybe the standard movie player has access to the oatv  huffman decoder.. (clutching at straws.)
it does not usethe meta file.. just tried  i the meta file the discription is displaying ok.. 
the meta file is recreated (but without description).
so i wonder whether it has access to the epgcache and pulls the data from there

**Comment by dazulrich** [comment]
Thanks @pjsharp. I may have one or two other odd samples that might be good to have analysed as.part.of your work. Instead of having whole bunch of encoded chars, it's only 1 or 2 at the end. And one eit actually crashes e2.

**Comment by pjsharp** [comment]
@dazulrich Please attach the files here. Thank you!

**Comment by dazulrich** [comment]
@pjsharp, here you go.
with your new code those all display ok!!!  no crash and no extra chars!
they probably are nothing special.. but with the original code they were not ok. 

the crash is similar to Tyri0n's , but differnt byte.. log extract inside the zip
[crash.zip](https://github.com/betonme/e2openplugin-EnhancedMovieCenter/files/10297251/crash.zip)
[extra_chars.zip](https://github.com/betonme/e2openplugin-EnhancedMovieCenter/files/10297252/extra_chars.zip)

**Comment by pjsharp** [comment]
@dazulrich I analyzed your files.
The errors with original code were caused by:
 - using the wrong length when parsing (therefore more characters than necessary)
- incorrect handling of the "default encoding", which could sometimes lead to a crash (if illegal characters occurred)

All these issues are already solved in my code (PR).

When I read your post:
> And one eit actually crashes e2.

I initially thought that the crash happened with my code. :)
